File: prime_triangle/prime_triangle.rs
fn is_prime(n: u32) -> bool {
    assert!(n < 64);
    ((1u64 << n) & 0x28208a20a08a28ac) != 0
}

fn prime_triangle_row(a: &mut [u32]) -> bool {
    if a.len() == 2 {
        return is_prime(a[0] + a[1]);
    }
    for i in 1..a.len() - 1 {
        if is_prime(a[0] + a[i]) {
            a.swap(i, 1);
            if prime_triangle_row(&mut a[1..]) {
                return true;
            }
            a.swap(i, 1);
        }
    }
    false
}

fn prime_triangle_count(a: &mut [u32], count: &mut u32) {
    if a.len() == 2 {
        if is_prime(a[0] + a[1]) {
            *count += 1;
        }
        return;
    }
    for i in 1..a.len() - 1 {
        if is_prime(a[0] + a[i]) {
            a.swap(i, 1);
            prime_triangle_count(&mut a[1..], count);
            a.swap(i, 1);
        }
    }
}

fn print(a: &[u32]) {
    if a.is_empty() {
        return;
    }
    print!("{:2}", a[0]);
    for x in &a[1..] {
        print!(" {:2}", x);
    }
    println!();
}

fn main() {
    use std::time::Instant;
    let start = Instant::now();
    for n in 2..21 {
        let mut a: Vec<u32> = (1..=n).collect();
        if prime_triangle_row(&mut a) {
            print(&a);
        }
    }
    println!();
    for n in 2..21 {
        let mut a: Vec<u32> = (1..=n).collect();
        let mut count = 0;
        prime_triangle_count(&mut a, &mut count);
        if n == 2 {
            print!("{count}");
        } else {
            print!(" {count}");
        }
    }
    println!();
    let time = start.elapsed();
    println!("\nElapsed time: {} milliseconds", time.as_millis());
}

File: prime_triangle/prime_triangle.swift
import Foundation

func isPrime(_ n: Int) -> Bool {
    guard n > 0 && n < 64 else {
        return false
    }
    return ((UInt64(1) << n) & 0x28208a20a08a28ac) != 0
}

func primeTriangleRow(_ a: inout ArraySlice<Int>) -> Bool {
    let start = a.startIndex
    let end = a.endIndex
    if a.count == 2 {
        return isPrime(a[start] + a[start + 1])
    }
    for i in start + 1..<end - 1 {
        if isPrime(a[start] + a[i]) {
            a.swapAt(i, start + 1)
            if primeTriangleRow(&a[start + 1..<end]) {
                return true
            }
            a.swapAt(i, start + 1)
        }
    }
    return false
}

func primeTriangleCount(_ a: inout ArraySlice<Int>) -> Int {
    let start = a.startIndex
    let end = a.endIndex
    var count = 0
    if a.count == 2 {
        if isPrime(a[start] + a[start + 1]) {
            count += 1
        }
    } else {
        for i in start + 1..<end - 1 {
            if isPrime(a[start] + a[i]) {
                a.swapAt(i, start + 1)
                count += primeTriangleCount(&a[start + 1..<end])
                a.swapAt(i, start + 1)
            }
        }
    }
    return count
}

func printRow(_ a: [Int]) {
    if a.count == 0 {
        return
    }
    print(String(format: "%2d", a[0]), terminator: "")
    for x in a[1...] {
        print(String(format: " %2d", x), terminator: "")
    }
    print()
}

for n in 2...20 {
    var a = Array(1...n)
    if primeTriangleRow(&a[...]) {
        printRow(a)
    }
}
print()

for n in 2...20 {
    var a = Array(1...n)
    if n > 2 {
        print(" ", terminator: "")
    }
    print("\(primeTriangleCount(&a[...]))", terminator: "")
}
print()
